docs/gps303.1
.TH GPS303 1 2022-05-27 "GPS303 Manipulation Tool" "User Commands"

.SH NAME

gps303 \- utility to send control messages to the GPS tracker terminal

.SH SYNOPSIS

gps303 [-c /path/to/config/file] [-d]

.SH DESCRIPTION

Command line tool to send "unsolicited reply" messages to the tracker.
There are messages to configure settings in the terminal, or to request
actions, such as to start a monitoring call, or to power off.

.SH OPTIONS

.TP 0.5i
.B -d
Set "debug" level of logging

.TP 0.5i
.B -c /path/to/config
Location of the configuration file (default /etc/gps303.conf)

.SH COMMANDS

After the options, IMEI (16 decimal characters) and the command verb
must be specified, optionally followed by a list of
.B key=value
pairs. The command verb is a case insensitive abbreviation that
uniquely identifies a class from the file gps303/gps303proto.py.
These classes correspond to message types recognized by the terminal.
Only a few of them are useful to construct "unsolicited replies".
.B key=value
pairs must specify kwargs for the constructor of the "Out" subclass
of the class.

.SH KNOWN BUGS

The command sends the message to the "collector" daemon for sending
to the terminal. There is no guarantee that the terminal is online
at the moment of execution, and there is no feedback. The only way
to verify that the command was successfully sent is to check the log
of the collector.

.SH COPYRIGHT

The program is copyrighted by Eugene G. Crosser <crosser@average.org>
and freely distributable under the terms of MIT license.

.SH CREDITS

The program is inspired by the project petGPS by Thomas Obadia

.SH SEE ALSO

.BR gps303.conf (5)


debian/changelog
gps303 (1.00) experimental; urgency=medium

  * test: include lookaside and termconfig in the loop
  * doc: edit and correct README.md
  * docs: expand man pages
  * forgotten manpage for install

 -- Eugene Crosser <crosser@average.org>  Thu, 23 Jun 2022 22:58:35 +0200

gps303 (0.99) experimental; urgency=medium

  * Revive mkgpx example script
  * Drop data if we are receiving junk
  * test: fuzzing collector
  * test: complete fuzzer unittest
  * collector: fix problems found by fuzzer test
  * test: make unittests individually runnable
  * collector: enable SO_KEEPALIVE on accepted conn.
  * test: message fuzzer and fix found problems
  * test: move `send_and_drain` to common module
  * test: skeleton for an end-to-end unittest
  * test: adjust sleeps and drains when daemons run
  * test: minimally functional test_storage
  * do not try to use zlib's _Decompress
  * test: use IPv4 because github actions
  * Add sample opencellid csv for tests
  * test: better aquisition of free ports
  * test: opencellid downloader

 -- Eugene Crosser <crosser@average.org>  Wed, 22 Jun 2022 18:04:10 +0200

gps303 (0.98) experimental; urgency=medium

  * include runtime deps as build-time for typecheck
  * fix l3str/l3int breakage provoked by typeckeck
  * make termconfig use unit specific config section
  * expand README: lookaside and termconfig sections.
  * wsgateway: reclassify http write hack as permanent
  * Update README with configuring the terminal
  * Implement remaining "Out" commands
  * do not run black older than 21.1

 -- Eugene Crosser <crosser@average.org>  Tue, 07 Jun 2022 00:17:55 +0200

gps303 (0.97) experimental; urgency=medium

  * adjust tests
  * typecheck: skip test if mypy verison < 0.942

 -- Eugene Crosser <crosser@average.org>  Tue, 31 May 2022 01:05:39 +0200

gps303 (0.96) experimental; urgency=medium

  * Do not write startup message for command-line cmds
  * Add a (short) man page
  * typing: make zmsg.py typecheck
  * typing: annotate gps303proto.py (mostly)
  * typechecking: less hacky deal with dynamic attrs
  * typing: annotate opencellid.py
  * typchecking: annotate googlemaps.py
  * typechecking: annotate evstore.py
  * typechecking: annotate common.py
  * typeckecing: annotate ocid_dload.py
  * typeckecing: annotate watch.py
  * typechecking: annotate termconfig.py
  * typeckecking: annotate storage.py
  * typing: annotate lookaside.py
  * typechecking: annotate collector.py
  * typeckeck: annotate __main__ and fix gps303proto
  * typeckecking: annotate wsgateway.py
  * formatting: revive black formatting
  * unittest: type checking and formatting

 -- Eugene Crosser <crosser@average.org>  Tue, 31 May 2022 00:36:33 +0200

gps303 (0.95) experimental; urgency=medium

  * Quick fix for a missing variable

 -- Eugene Crosser <crosser@average.org>  Thu, 26 May 2022 19:32:44 +0200

gps303 (0.94) experimental; urgency=medium

  * Opencellid download service
  * Improve error handling of the downloader
  * cleanup and document config file a bit
  * Prepare to have dynamically chosen lookup backend
  * Add googlemaps lookaside backend

 -- Eugene Crosser <crosser@average.org>  Thu, 26 May 2022 19:25:40 +0200

gps303 (0.93) experimental; urgency=medium

  * use fixed github action
  * try to make github actions work
  * "the `uses' attribute must be a path, a Docker image, or owner/repo@ref"
  * Update README
  * actions: action-gh-release, not artifact upload
  * Beautify logging
  * Make a `make-release` script to run `gbp dch`

 -- Eugene Crosser <crosser@average.org>  Tue, 24 May 2022 22:53:13 +0200

gps303 (0.92) experimental; urgency=low

  [ Eugene Crosser ]
  * "When you start a pattern with *, you must use quotes."
  * add github action
  * allow imei as the document path
  * Report status (with battery %) to the webpage
  * Add references to the doc/source
  * webdemo: draw very simplistic line between locs

 -- Eugene Crosser <crosser@average.org>  Mon, 23 May 2022 23:55:46 +0200

gps303 (0.91) experimental; urgency=low

  [ Eugene Crosser ]
  * fix log message using unknown variable

 -- Eugene Crosser <crosser@average.org>  Tue, 10 May 2022 09:42:30 +0200

gps303 (0.90) experimental; urgency=low

  [ Eugene Crosser ]
  * Expand README
  * clean a couple of TODOs in wsgateway
  * make parse_message return UNKNOWN on parse crash
  * reimplement backlog query again
  * A hack in packet framing to false stop bytes match
  * do not respond to hibernation; minor cleanup
  * wsgateway with new message distribution model
  * fix zmq subscription topics
  * misc black reformatting
  * fill in `when` in Resp packet
  * WIP retoure messaging
  * Partly revert "Broadcast location, gps and approximated"

 -- Eugene Crosser <crosser@average.org>  Tue, 10 May 2022 09:09:08 +0200

gps303 (0.01) experimental; urgency=low

  [ Eugene Crosser ]
  * adjust debianization to wsgateway
  * Update evstore schema to support in and out msgs
  * WIP on supporting multiple markers
  * WIP on baclog fetcher
  * lookaside - do not look for offline positioning
  * wsgateway aggregate subscriptions upstream
  * support subscriptions on the client side
  * wsgateway read html file every time
  * Webdemo use its url as the base for ws url
  * wsgateway properly handle write-busy websockets
  * Make constructor of protocol objects more elegant
  * mock use persistent history for enrered commands
  * Show coordinates on the map
  * WIP on ws gateway, it now works
  * Make ws gateway act as http server too
  * WIP on websocket gateway
  * Try to make systemd target more robust
  * lintian: override "unusual target"
  * some cleanup of debianization
  * set umask for group-writable unix domain sockets
  * Change LocEvt to use json encoding
  * remove local conf, add license
  * Broadcast location, gps and approximated
  * clean up `from_packet()`
  * expand README
  * update gitignore debianization
  * debianize
  * add command line script
  * cleanup of gps303proto

gps303.py
#!/usr/bin/env python3

from enum import Enum
from select import poll, POLLIN, POLLERR, POLLHUP, POLLPRI
from socket import socket, AF_INET, SOCK_STREAM, SOL_SOCKET, SO_REUSEADDR
from struct import pack, unpack
import sys
from time import time

PORT = 4303

class P(Enum):
    UNKNOWN = 0x00
    LOGIN = 0x01
    STATUS = 0x13
    HIBERNATION = 0x14
    time = 0x30
    SETUP = 0x57
    


def answer_setup(data):
    return bytes.fromhex("0300310000000000000000000000000000000000000000000000003b3b3b")

def handle_packet(packet, addr, when):
    xx, length, proto = unpack("!2sBB", packet[:4])
    crlf = packet[-2:]
    data = packet[4:-2]
    if xx != b"xx" or crlf != b"\r\n" or (length > 1 and len(packet) != length + 2):
        print("bad packet:", packet.hex())
    print("length", length, "proto", hex(proto))
    try:
        p = P(proto)
    except ValueError:
        p = P.UNKNOWN
    payload = b""
    if p == P.LOGIN:
    	print("imei", data[:-1].hex(), "ver", data[-1:].hex())
    elif p == P.SETUP:
        payload = answer_setup(data)
    length = len(payload)+1
    if length > 6:
        length -= 6
    return b"xx" + pack("B", length) + pack("B", proto) + payload + b"\r\n"

if __name__ == "__main__":
    ctlsock = socket(AF_INET, SOCK_STREAM)
    ctlsock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
    ctlsock.bind(("", PORT))
    ctlsock.listen(5)
    ctlfd = ctlsock.fileno()
    pollset = poll()
    pollset.register(ctlfd, POLLIN | POLLERR | POLLHUP | POLLPRI)
    clnt_dict = {}
    while True:
        try:
            events = pollset.poll(1000)
        except KeyboardInterrupt:
            print("Exiting")
            sys.exit(0)
        for fd, ev in events:
            if fd == ctlfd:
                if ev & POLLIN:
                    clntsock, clntaddr = ctlsock.accept()
                    clntfd = clntsock.fileno()
                    clnt_dict[clntfd] = (clntsock, clntaddr)
                    pollset.register(clntfd, POLLIN | POLLERR | POLLHUP | POLLPRI)
                    print("accepted connection from", clntaddr, "fd", clntfd)
                if ev & ~POLLIN:
                    print("unexpected event on ctlfd:", ev)
            else:
                try:
                    clntsock, clntaddr = clnt_dict[fd]
                except KeyError:  # this socket closed already
                    continue
                if ev & POLLIN:
                    packet = clntsock.recv(4096)
                    when = time()
                    print("packet", packet, "from", clntaddr, "from fd", fd)
                    if packet:
                        response = handle_packet(packet, clntaddr, when)
                        if response:
                            try:
                                # Ignore possibility of blocking
                                clntsock.send(response)
                            except OSError as e:
                                print("sending to socket", fd, "error", e)
                    else:
                        print("disconnect")
                        pollset.unregister(fd)
                        clntsock.close()
                        del clnt_dict[fd]
                if ev & ~POLLIN:
                    print("unexpected event", ev, "on fd", fd)
